classes/access_manager.php

```diff
@@ -0,0 +1,147 @@
+
+    /** @var quiz $quiz A quiz object containing all information pertaining to current quiz. */
+    private $quiz;
+
+    /** @var quiz_settings $quizsettings A quiz settings persistent object containing plugin settings */
+    private $quizsettings;
+
+    /** @var context_module $context Context of this quiz activity. */
+    private $context;
+
+    /**
+     * The access_manager constructor.
+     *
+     * @param quiz $quiz The details of the quiz.
+     *
+     * @throws \moodle_exception
+     */
+    public function __construct(quiz $quiz) {
+        $this->quiz = $quiz;
+        $this->context = context_module::instance($quiz->get_cmid());
+        $this->quizsettings = quiz_settings::get_record(['quizid' => $quiz->get_quizid()]);
+        if ($this->quizsettings == false) {
+            throw new \moodle_exception('noconfigfound', 'quizaccess_seb', '', $quiz->get_cmid());
+        }
+    }
+
+    /**
+     * Check if the current request contains the required key hashes in header.
+     *
+     * @return bool true if the user is using a browser with a permitted key, false if not.
+     *
+     * @throws \coding_exception
+     */
+    public function validate_access_keys() : bool {
+        $configkey = $this->quizsettings->get('configkey');
+        if (empty($configkey)) {
+            return false;
+        }
+
+        // If the Config Key header isn't present, prevent access.
+        if (!array_key_exists(self::CONFIG_KEY_HEADER, $_SERVER)) {
+            return false;
+        }
+
+        return self::check_key($configkey, self::get_this_page_url(),
+                trim($_SERVER[self::CONFIG_KEY_HEADER]));
+    }
+
+    /**
+     * Check if Safe Exam Browser is required to access quiz.
+     *
+     * @return bool If required.
+     *
+     * @throws \coding_exception
+     */
+    public function seb_required() : bool {
+        return $this->quizsettings->get('requiresafeexambrowser') == true;
+    }
+
+    /**
+     * Check if user has any capability to bypass the Safe Exam Browser requirement.
+     *
+     * @return bool True if user can bypass check.
+     *
+     * @throws \coding_exception
+     */
+    public function can_bypass_seb() : bool {
+        return has_capability('quizaccess/seb:bypassseb', $this->context);
+    }
+
+    /**
+     * Return the full URL that was used to request the current page, which is
+     * what we need for verifying the X-SafeExamBrowser-RequestHash header.
+     */
+    private function get_this_page_url() : string {
+        global $FULLME;
+        return $FULLME;
+    }
+
+    /**
+     * Check the hash from the request header against the permitted keys.
+     *
+     * @param array $keys allowed keys.
+     * @param string $url the request URL.
+     * @param string $header the value of the X-SafeExamBrowser-RequestHash to check.
+     * @return bool true if the hash matches.
+     */
+    private function check_keys(array $keys, $url, $header) : bool {
```

Comment:
should we cast strings for url and header?

Reply:
is there any chance that url will be different to $FULLME? Maybe we can set $this->url in constructor? We can have it as a second parameter in constructor for testing and if not set then fall back to  $FULLME

Reply:
Copy-paste error for casting. Fixed.

I'm not sure how the rule objects are reused between different pages so hesitant to hard set it in the constructor. However, I added an optional param to the public method `validate_access_keys` for pageurl that falls back to $FULLME if empty. Hopefully this will cover most use cases.